File: php/DataHub.php
<?php
use Workerman\Connection\TcpConnection;
require_once __DIR__ . '/RandomDataGenerator.php';
class DataHub
{
    private array $connections = [];
    private int $rate;
    private RandomDataGenerator $randomDataGenerator;
    public function __construct(int $rate)
    {
        $this->rate = $rate;
        $this->randomDataGenerator = new RandomDataGenerator();
    }

    public function addConnection(TcpConnection $connection)
    {
        echo "New Connection added, id: $connection->id\n";
        $this->connections[] = $connection;
        if(count($this->connections) == 1) {
            // start data when at least one connection is connected
            $this->generateAndSendData();
        }
    }

    public function removeConnection(TcpConnection $connection)
    {
        $index = array_search($connection, $this->connections);
        if ($index !== false) {
            echo "Connection removed, id: $connection->id\n";
            unset($this->connections[$index]);
        }
        if(count($this->connections) == 0) {
            echo "No more connections, stop generating data\n";
            \Workerman\Lib\Timer::delAll();
        }
    }

    public function generateAndSendData() {
        $timeout_between_data = 60 / $this->rate;
        echo "Start data generation";
        \Workerman\Lib\Timer::add($timeout_between_data, function()
        {
            $data = $this->randomDataGenerator->getData();
            echo "Data: $data\n";
            foreach ($this->connections as $connection) {
                $connection->send($data);
            }
        });
    }
}

File: cpp/src/DataHub.cpp
#include "DataHub.h"
#include <iostream>
#include <chrono>
#include <thread>

DataHub::DataHub(int rate, WebsocketServer* server) 
    : rate(rate), 
      randomDataGenerator(),
      isRunning(false),
      serverPtr(server) {
}

DataHub::~DataHub() {
    stopBroadcasting();
}

void DataHub::addConnection(ConnectionHandle hdl) {
    std::lock_guard<std::mutex> lock(connectionsMutex);
    std::cout << "New connection added" << std::endl;
    connections.push_back(hdl);
    
    if (connections.size() == 1) {
        // Start broadcasting when first connection is made
        startBroadcasting();
    }
}

void DataHub::removeConnection(ConnectionHandle hdl) {
    std::lock_guard<std::mutex> lock(connectionsMutex);
    
    // Find and remove the connection
    connections.erase(
        std::remove_if(connections.begin(), connections.end(),
            [&](const ConnectionHandle& conn) {
                return !serverPtr->get_con_from_hdl(conn)->get_raw_socket().is_open() ||
                       serverPtr->get_con_from_hdl(hdl).get() == serverPtr->get_con_from_hdl(conn).get();
            }),
        connections.end());
    
    std::cout << "Connection removed, remaining: " << connections.size() << std::endl;
    
    if (connections.empty()) {
        stopBroadcasting();
    }
}

void DataHub::broadcastData() {
    while (isRunning) {
        std::string data = randomDataGenerator.getData();
        std::cout << "Data: " << data << std::endl;
        
        // Calculate sleep time between messages based on rate
        int sleepTimeMs = static_cast<int>(60.0 / rate * 1000);
        
        {
            std::lock_guard<std::mutex> lock(connectionsMutex);
            // Remove closed connections
            connections.erase(
                std::remove_if(connections.begin(), connections.end(),
                    [this](const ConnectionHandle& hdl) {
                        return !serverPtr->get_con_from_hdl(hdl)->get_raw_socket().is_open();
                    }),
                connections.end());
                
            // Send data to all active connections
            for (auto& conn : connections) {
                if (serverPtr->get_con_from_hdl(conn)->get_raw_socket().is_open()) {
                    try {
                        serverPtr->send(conn, data, websocketpp::frame::opcode::text);
                    } catch (const std::exception& e) {
                        std::cerr << "Error sending message: " << e.what() << std::endl;
                    }
                }
            }
        }
        
        // Sleep until next message
        std::this_thread::sleep_for(std::chrono::milliseconds(sleepTimeMs));
    }
}

void DataHub::startBroadcasting() {
    if (!isRunning) {
        isRunning = true;
        std::cout << "Start data generation" << std::endl;
        broadcastThread = std::thread(&DataHub::broadcastData, this);
    }
}

void DataHub::stopBroadcasting() {
    if (isRunning) {
        std::cout << "No more connections, stop generating data" << std::endl;
        isRunning = false;
        if (broadcastThread.joinable()) {
            broadcastThread.join();
        }
    }
}
